PlayerSkeleton.java

```diff
@@ -40,10 +40,89 @@ private static double evaluate(SerializedState s, double[] weights) {
 		valuation += s.cleared * weights[0];
 		valuation += getHighestCol(s.field) * weights[1];
 		valuation += getLowestCol(s.field) * weights[2];
+		valuation +=  getHolesCount(s.field, s.tops) * weights[3];
+		valuation += getBlockadeCount(s.field, s.tops) * weights[4];
+		valuation += getParityCount(s.field) * weights[5] ;
+		valuation += getTotalHeightDiff(s.tops) * weights[6];
 
 		return valuation;
 	}
 
+	private static int getHolesCount(int[][] field, int[] tops) {
+
+		int holeCount = 0;
+		boolean isHole = false;
+		for (int i = 0; i < State.COLS; i++) {
+			for (int j = tops[i]; j >= 0; j--) {
+			    if (field[j][i] != 0) {
+			        isHole = true;
+                } else if(isHole && field[j][i] == 0) {
+			        holeCount++;
+			        isHole = false;
+                }
+            }
+            isHole = false;
+		}
+
+		return holeCount;
+	}
+
+	private static int getBlockadeCount (int[][] field, int[] tops) {
+
+	    int totalBlockade = 0;
+	    int tempBlockade = 0;
+	    boolean isHole = false;
+	    for (int i = 0; i < State.COLS; i++) {
+	        for (int j = tops[i]; j >= 0; j--) {
+	            if (field[j][i] != 0) {
+	                isHole = true;
+	                tempBlockade++;
+                } else if (isHole) {
+	                totalBlockade += tempBlockade;
+	                tempBlockade = 0;
+	                isHole = false;
+                }
+            }
+            //reset tempBlockade when changing cols
+            tempBlockade = 0;
+        }
+        return totalBlockade;
+    }
+
+    private static int getParityCount (int[][] field) {

```

Comment:
I meant parity as in when you imagine the field as a checkered board, the `abs(number of whites - number of blacks)`.